File: src/kernelprolog/fluents/SxxMachine/IterableSource.cpp
using namespace std;

#include "IterableSource.h"
#include "../../terms/SxxMachine/Prog.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../main/SxxMachine/Builtins.h"

namespace SxxMachine
{

	IterableSource::IterableSource(Prog *p) : Source(p)
	{
		e = nullptr;
	}

	IterableSource::IterableSource(Iterator *iterator, Prog *p) : Source(p)
	{
		this->e = iterator;
	}

	IterableSource::IterableSource(vector V, Prog *p) : Source(p)
	{
		this->e = V.begin();
	}

	Term *IterableSource::getElement()
	{
		if (nullptr == e || !e->hasNext())
		{
			return nullptr;
		}
		else
		{
			return static_cast<Term*>(e->next());
		}
	}

	void IterableSource::stop()
	{
		e = nullptr;
	}
}

File: jsrc-kernelprolog/fluents/SxxMachine/IterableSource.java
package SxxMachine;

import java.util.Iterator;

import SxxMachine.pterm.SourceFluentTerm;

/**
 * Builds Prolog Iterators from Java Sequences and Iterator type classes
 */
public class IterableSource extends SourceFluentTerm {
    private Iterator e;

    public IterableSource(Prog p) {
        super(p);
        e = null;
    }

    public IterableSource(Iterator iterator, Prog p) {
        super(p);
        this.e = iterator;
    }
 

    public IterableSource(Iterable V, Prog p) {
        super(p);
        this.e = V.iterator();
    }

    @Override
    public Term getElement() {
        if (null == e || !e.hasNext())
            return null;
        else
            return (Term) e.next();
    }

    @Override
    public void stop() {
        e = null;
    }
}
